clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DelegatingAvroStoreClient.java

```diff
@@ -12,8 +13,41 @@
 /**
  * Inside Fast-Client, we choose to use n-tier architecture style to build a pipeline to separate different
  * types of logic in different layer.
- * We should follow this pattern if we don't have any strong concerns in the future development.
+ *
+ * n-tier architecture => Having multiple layers where each layer wraps the next inner one.
+ * Each layer provides some functionality, e.g. stats collection, etc.
+ *
+ * Fast-Client's layers include the below components. Check {@link ClientFactory#getAndStartGenericStoreClient}
+ * to figure out how the layers are put together for different requirements.
+ *
+ * Layer -1:
+ * AvroGenericStoreClient => interface: Borrowed from thin-client
+ *
+ * Layer 0:
+ * InternalAvroStoreClient implements AvroGenericStoreClient => Only 1 abstract class implementing above interface:
+ *                            Both thin-client and fast-client uses this as the Layer 0. All the internal
+ *                            implementations of different tiers should extend this class.
+ *
+ * Layer 1:
+ * DispatchingAvroGenericStoreClient extends InternalAvroStoreClient => in charge of routing and serialization/de-serialization
+ *
+ * Layer 2:
+ * RetriableAvroGenericStoreClient extends DelegatingAvroStoreClient => Adds optional retry ability on top of DispatchingAvroGenericStoreClient
+ *
+ * Layer 3:
+ * StatsAvroGenericStoreClient extends DelegatingAvroStoreClient => Adds stats on top of Layer 2 or Layer 1. There is no option
+ *                            to disable it, but if needed, can be disabled.
+ *
+ * Layer 4:
+ * DualReadAvroGenericStoreClient extends DelegatingAvroStoreClient => Adds an extra read via thin client on top of Layer 3.
+ *
+ * utils class:
+ * DelegatingAvroStoreClient extends InternalAvroStoreClient => Delegator pattern to not override all the
+ *                           functions in every superclass in a duplicate manner.
+ *
+ * Interactions between these class for some flows: https://swimlanes.io/u/D3E9Q50pb
```

Comment:
Is this link accessible permanently?
If not, we should remove it and if you are interested, maybe we could document some ASCII-based workflow, and in the high-level, I feel the above javadoc is clear enough.

Reply:
You could attach a digram to the PR as well and in github docs.

Reply:
Yes, It's accessible permanently. I attached the link as this can be improved continuously.